cmake/Install.cmake

```diff
@@ -9,4 +9,107 @@
 if(NOT INSTALLED_ENGINE)
     ly_get_absolute_pal_filename(pal_dir ${CMAKE_CURRENT_SOURCE_DIR}/cmake/Platform/${PAL_PLATFORM_NAME})
     include(${pal_dir}/Install_${PAL_PLATFORM_NAME_LOWERCASE}.cmake)
-endif()
\ No newline at end of file
+endif()
+
+#! ly_install_directory: specifies a directory to be copied to the install layout at install time
+#
+# \arg:DIRECTORIES directories to install
+# \arg:DESTINATION (optional) destination to install the directory to (relative to CMAKE_PREFIX_PATH)
+# \arg:EXCLUDE_PATTERNS (optional) patterns to exclude
+#
+# \notes: refer to cmake's install(DIRECTORY documentation for more information
+#
+function(ly_install_directory)
+
+    set(options)
+    set(oneValueArgs DESTINATION)
+    set(multiValueArgs DIRECTORIES EXCLUDE_PATTERNS)
+
+    cmake_parse_arguments(ly_install_directory "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+    if(NOT ly_install_directory_DIRECTORIES)
+        message(FATAL_ERROR "You must provide at least a directory to install")
+    endif()
+
+    foreach(directory ${ly_install_directory_DIRECTORIES})
+
+        cmake_path(ABSOLUTE_PATH directory)
+
+        if(NOT ly_install_directory_DESTINATION)
+            # maintain the same structure relative to LY_ROOT_FOLDER
+            set(ly_install_directory_DESTINATION ${directory})
+            if(${ly_install_directory_DESTINATION} STREQUAL ".")
+                set(ly_install_directory_DESTINATION ${CMAKE_CURRENT_LIST_DIR})
+            else()
+                cmake_path(ABSOLUTE_PATH ly_install_directory_DESTINATION BASE_DIRECTORY ${CMAKE_CURRENT_LIST_DIR})
+            endif()
+            # take out the last directory since install asks for the destination of the folder, without including the fodler itself
+            cmake_path(GET ly_install_directory_DESTINATION PARENT_PATH ly_install_directory_DESTINATION)
+            cmake_path(RELATIVE_PATH ly_install_directory_DESTINATION BASE_DIRECTORY ${LY_ROOT_FOLDER})       
+        endif()
```

Comment:
We should add a check after this `if` block to make sure the LY_ROOT_FOLDER is a prefix path of the DESTINATION and FATAL_ERROR if not.

```cmake
cmake_path(IS_PREFIX LY_ROOT_FOLDER ${ly_install_directory_DESTINATION} is_destination_in_engine)
if(NOT is_destination_in_engine)
    message(FATAL_ERROR "The Destination directory \"${ly_install_directory_DESTINATION}\"must be within the LY_ROOT_FOLDER of \"${LY_ROOT_FOLDER}\"")
endif()
```

We should also add this check to `ly_install_files` command as well.

Reply:
the `DESTINATION` directory ends up being relative to `CMAKE_PREFIX_PATH`, here we are assuming that it will take the same path relative to the `LY_ROOT_FOLDER` if not specified. But the final path is not under `LY_ROOT_FOLDER`

Reply:
But if the user specifies an absolute path to the DESTINATION that is outside of the LY_ROOT_FOLDER, then the `cmake_path(RELATIVE_PATH)` doesn't modify the DESTINATION.